ouroboros-consensus/src/ouroboros-consensus/Ouroboros/Consensus/MiniProtocol/ChainSync/Client/HistoricityCheck.hs

```diff
@@ -0,0 +1,163 @@
+{-# LANGUAGE DeriveAnyClass      #-}
+{-# LANGUAGE DerivingStrategies  #-}
+{-# LANGUAGE GADTs               #-}
+{-# LANGUAGE LambdaCase          #-}
+{-# LANGUAGE NamedFieldPuns      #-}
+{-# LANGUAGE ScopedTypeVariables #-}
+{-# LANGUAGE StandaloneDeriving  #-}
+{-# LANGUAGE TypeApplications    #-}
+
+module Ouroboros.Consensus.MiniProtocol.ChainSync.Client.HistoricityCheck (
+    -- * Interface
+    HistoricalChainSyncMessage (..)
+  , HistoricityCheck (..)
+  , HistoricityCutoff (..)
+  , HistoricityException (..)
+    -- * Real implementation
+  , mkCheck
+  , noCheck
+  ) where
+
+import           Control.Exception (Exception)
+import           Control.Monad (when)
+import           Control.Monad.Except (throwError)
+import           Data.Time.Clock (NominalDiffTime)
+import           Data.Typeable (eqT)
+import           Ouroboros.Consensus.Block
+import           Ouroboros.Consensus.BlockchainTime (RelativeTime,
+                     SystemTime (..), diffRelTime)
+import           Ouroboros.Consensus.HeaderStateHistory
+                     (HeaderStateWithTime (..))
+import           Ouroboros.Consensus.HeaderValidation (HasAnnTip,
+                     headerStatePoint)
+import           Ouroboros.Consensus.Node.GsmState (GsmState (..))
+
+{-------------------------------------------------------------------------------
+  Interface
+-------------------------------------------------------------------------------}
+
+-- | Interface for the ChainSync client for deciding whether @MsgRollBackward@s
+-- and @MsgAwaitReply@s are historical.
+data HistoricityCheck m blk = HistoricityCheck {
+    -- | Determine whether the received @MsgRollBackward@ is historical. The
+    -- input is the oldest state that was rolled back. (Note that rollbacks of
+    -- depth zero are hence never historical.).
+    judgeMessageHistoricity ::
+         HistoricalChainSyncMessage
+      -> HeaderStateWithTime blk
+      -> m (Either HistoricityException ())
+  }
+
+-- | ChainSync historicity checks are performed for @MsgRollBackward@s and
+-- @MsgAwaitReply@s, see 'HistoricityCheck'.
+data HistoricalChainSyncMessage =
+    HistoricalMsgRollBackward
+  | HistoricalMsgAwaitReply
+  deriving stock (Show, Eq)
+
+data HistoricityException =
+  -- | We received a @MsgRollBackward@ or a @MsgAwaitReply@ at a time too early
+  -- for it to plausibly have been sent by an honest caught-up peer.
+  --
+  -- INVARIANT: @'historicityCutoff' < 'arrivalTime' `diffRelTime` 'slotTime'@
+  forall blk. HasHeader blk => HistoricityException {
+      historicalMessage :: HistoricalChainSyncMessage
+      -- | Depending on 'historicalMessage':
+      --
+      --  * 'HistoricalMsgRollBackward': The oldest header that was rewound.
+      --
+      --  * 'HistoricalMsgAwaitReply': The tip of the candidate fragment.
+    , historicalPoint   :: !(Point blk)
+      -- | The time corresponding to the slot of the oldest header that was
```

Comment:
This comment's current wording doesn't work for `MsgAwaitReply`.

Reply:
Changed to
```haskell
      -- | The time corresponding to the slot of 'historicalPoint'.
```